### PR TITLE
Refactor application version filtering

### PR DESCRIPTION
**Pull request by diitaz93**
### Added

- Filter query functions for Application Version
- Test functions for the filter functions
- Store fixtures for the test functions

### Changed

- Reorganized functions and tests so that they are in alphabetical order of their respective model

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b refactor-app-version -a
    ```

### How to test

- [x] Add a new sample with: (any existing application tag and customer should work, any new sample name should work)
```
cg add sample cust000 sample_1 -s male -a WGSPCFC030
``` 
- [x] Verify the information of the sample with: (match the sample name from the previous step output)
```
cg get sample overlysuitedsponge
``` 
- [x] Change the application tag of the sample with: 
```
cg set sample overlysuitedsponge -kv application_version WGSPCFC030 --skip-lims -y
```

### Expected test outcome

- [x] The sample was added correctly with the correct application tag
- [x] The sample appreciation tag could be changed successfully

## Review

- [x] Tests executed by SD
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by diitaz93** [comment]
Not ready but would be good some input from now

File: cg/store/api/find_basic_data.py [hunk 1]
```diff
@@ -107,9 +112,16 @@ def get_applications(self) -> List[Application]:
         )
 
     def application_version(self, application: Application, version: int) -> ApplicationVersion:
```
Comment (diitaz93) [inline cg/store/api/find_basic_data.py:114]: renaming this to `get_application_version` will be another PR

File: cg/store/api/find_basic_data.py [hunk 2]
```diff
@@ -159,23 +171,26 @@ def get_collaboration_by_internal_id(self, internal_id: str) -> Collaboration:
         ).first()
 
     def current_application_version(self, tag: str) -> Optional[ApplicationVersion]:
```
Comment (diitaz93) [inline cg/store/api/find_basic_data.py:173]: renaming this function will be included in another pr

File: tests/store/api/find/test_find_basic_data.py [hunk 3]
```diff
@@ -156,6 +165,26 @@ def test_get_applications_by_prep_category_and_is_not_archived(
...
+    assert isinstance(application, Application)
+
+    # WHEN calling the function with an application and a version
+    application_version: ApplicationVersion = local_store.application_version(
+        application=application,
+        version=application_version_version,
+    )
+
+    # THEN
+    assert isinstance(application_version, ApplicationVersion)
```
Comment (ChrOertlin) [inline tests/store/api/find/test_find_basic_data.py:185]:
```suggestion
def test_application_version_with_version_and_application(
    store_with_applications_with_application_versions: Store,
    application_version_version: int,
):
    """Test that application_version returns a valid query given an application and a version"""
    # GIVEN a store with applications and an ApplicationVersion.version

    application: Application = store_with_applications_with_application_versions.get_applications()[0]
    assert isinstance(application, Application)

    # WHEN calling the function with an application and a version
    application_version: ApplicationVersion = store_with_applications_with_application_versions.application_version(
        application=application,
        version=application_version_version,
    )

    # THEN
    assert isinstance(application_version, ApplicationVersion)
```

File: tests/store/conftest.py [hunk 4]
```diff
@@ -264,6 +264,35 @@ def fixture_store_with_an_application_with_and_without_attributes(
...
+    """Returns a store with two different applications and their respective application versions."""
+    applications: List[
+        Application
+    ] = store_with_an_application_with_and_without_attributes.get_applications()
+
+    for app in applications:
+        helpers.ensure_application_version(
+            store=store_with_an_application_with_and_without_attributes,
+            application_tag=app.tag,
+        )
```
Comment (ChrOertlin) [inline tests/store/conftest.py:292]:
```suggestion
    for application in applications:
        helpers.ensure_application_version(
            store=store_with_an_application_with_and_without_attributes,
            application_tag=application.tag,
        )
```

Comment (ChrOertlin) [reply tests/store/conftest.py:292]: In other parts of the code we spell out the variables.

File: tests/store/filters/test_status_application_version_filters.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    helpers: StoreHelpers,
+):
+    """Test that the application of the filtered application version is the correct one."""
+    # GIVEN a store with different applications
+    applications: List[
+        Application
+    ] = store_with_applications_with_application_versions.get_applications()
+    linked_application: Application = applications[0]
+    unlinked_application: Application = applications[-1]
+    assert str(unlinked_application) != str(linked_application)
```
Comment (ChrOertlin) [inline tests/store/filters/test_status_application_version_filters.py:33]: is the `str()` here needed?  They should both be of type application or am I missing something?
Same for the others below.

**Review by ChrOertlin — COMMENTED** [review]
Looks good so far!

File: cg/store/api/find_basic_data.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 from sqlalchemy import desc
 from sqlalchemy.orm import Query
+from sqlalchemy.types import DateTime
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:7]: done

File: cg/store/filters/status_application_version_filters.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+def filter_application_versions_before_date(
+    application_versions: Query, date: DateTime, **kwargs
+) -> Query:
+    """Return the application versions with valid_from before a given date."""
+    return application_versions.filter(ApplicationVersion.valid_from < date)
+
+
+def order_application_versions_by_desc_date(application_versions: Query, **kwargs) -> Query:
+    """Returned the application versions ordered by valid_from in descending order."""
+    return application_versions.order_by(desc(ApplicationVersion.valid_from))
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/store/filters/status_application_version_filters.py:40]: renamed 👍

File: tests/store/filters/test_status_application_version_filters.py (hunk 5 above)
Comment (diitaz93) [reply tests/store/filters/test_status_application_version_filters.py:34]: not really, deleting

**Comment by diitaz93** [comment]
This is ready now @seallard @ChrOertlin @henrikstranneheim

**Review by ChrOertlin — APPROVED** [review]
looks good. I made some comments and suggestions

File: cg/store/api/find_basic_data.py [hunk 8]
```diff
@@ -106,10 +109,21 @@ def get_applications(self) -> List[Application]:
...
+
     def application_version(self, application: Application, version: int) -> ApplicationVersion:
-        """Fetch an application version."""
-        query = self.ApplicationVersion.query.filter_by(application=application, version=version)
-        return query.first()
+        """Return an application version."""
+        return apply_application_versions_filter(
+            application_versions=self._get_query(table=ApplicationVersion),
+            application=application,
+            version=version,
```
Comment (seallard) [inline cg/store/api/find_basic_data.py:121]: It might be better to only send in the property we are filtering on instead of the entire object.

Comment (seallard) [reply cg/store/api/find_basic_data.py:121]: Using only the IDs simplifies the filter functions and reduces the amount of data being passed around

File: cg/store/api/find_basic_data.py [hunk 9]
```diff
@@ -159,23 +173,26 @@ def get_collaboration_by_internal_id(self, internal_id: str) -> Collaboration:
...
+        application_id = application.id
+        return apply_application_versions_filter(
+            filter_functions=[
+                ApplicationVersionFilter.FILTER_BY_APPLICATION_ID,
+                ApplicationVersionFilter.FILTER_BY_DATE,
+                ApplicationVersionFilter.ORDER_BY_VALID_FROM,
+            ],
+            application_versions=self._get_query(table=ApplicationVersion),
+            application_id=application_id,
+            date=dt.datetime.now(),
```
Comment (seallard) [inline cg/store/api/find_basic_data.py:189]: Can we really always filter on the current time stamp? Will it find a matching application version? Maybe the parameter can have a more descriptive name. We probably want separate date parameters for FILTER_BY_DATE and ORDER_BY_VALID_FROM

File: cg/store/api/find_basic_data.py [hunk 10]
```diff
@@ -159,23 +173,26 @@ def get_collaboration_by_internal_id(self, internal_id: str) -> Collaboration:
...
+        ).first()
 
     def latest_version(self, tag: str) -> Optional[ApplicationVersion]:
         """Fetch the latest application version for an application tag."""
-        application_obj = self.Application.query.filter_by(tag=tag).first()
-        return (
-            application_obj.versions[-1] if application_obj and application_obj.versions else None
-        )
+        application = self.get_application_by_tag(tag=tag)
+        return application.versions[-1] if application and application.versions else None
```
Comment (seallard) [inline cg/store/api/find_basic_data.py:195]: What happens if the versions list is empty? Add a test case and maybe some error handling

File: cg/store/api/find_basic_data.py (hunk 8 above)
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:121]: Included in another PR

File: cg/store/api/find_basic_data.py (hunk 9 above)
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:189]: agree! implemented

File: cg/store/api/find_basic_data.py (hunk 10 above)
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:195]: will include in another PR after refactoring this function

File: cg/store/api/find_basic_data.py [hunk 11]
```diff
@@ -159,23 +171,26 @@ def get_collaboration_by_internal_id(self, internal_id: str) -> Collaboration:
...
+                ApplicationVersionFilter.FILTER_BY_APPLICATION_ID,
+                ApplicationVersionFilter.FILTER_BY_DATE,
+                ApplicationVersionFilter.ORDER_BY_VALID_FROM,
+            ],
+            application_versions=self._get_query(table=ApplicationVersion),
+            application_id=application_id,
+            date=dt.datetime.now(),
+        ).first()
 
     def latest_version(self, tag: str) -> Optional[ApplicationVersion]:
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:192]: #1850

File: cg/store/api/find_basic_data.py (hunk 2 above)
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:175]: #1850

File: cg/store/api/find_basic_data.py (hunk 1 above)
Comment (diitaz93) [reply cg/store/api/find_basic_data.py:116]: #1850

File: tests/store/filters/test_status_application_version_filters.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+        table=ApplicationVersion
+    )
+    assert app_version_query.first().application == linked_application
+
+    # WHEN filtering an application version query by the linked application
+    filtered_app_version_query: Query = filter_application_versions_by_application(
+        application_versions=app_version_query,
+        application=linked_application,
+    )
+
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply tests/store/filters/test_status_application_version_filters.py:46]: This assertion is covered by the last assertion: If `filtered_app_version_query` is not a valid query it will not be able to call the `.all()` function and an error would be raised

File: tests/store_helpers.py [hunk 13]
```diff
@@ -75,17 +75,18 @@ def ensure_application_version(
         sequencing_depth: int = None,
         is_accredited: bool = False,
         version: int = 1,
+        valid_from: datetime = datetime.now(),
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply tests/store_helpers.py:78]: can't use this here due to circular imports
